src/Compiler/TypedTree/TypeProviders.fs

```diff
@@ -625,6 +630,36 @@ type ProvidedAssembly (x: Assembly) =
 
     override _.GetHashCode() = assert false; x.GetHashCode()
 
+[<NoComparison; NoEquality>]
+type CrackedStaticArgument = {
+    Name: string
+    Value: obj
+    ValueRange: range
+    /// Range stripped of quotes if the argument is a string constant
+    ValueRangeAdjusted: range option
+}
+
+type TypeProviderDiagnosticsContext (staticArgs: CrackedStaticArgument[], tpDesignation, fallbackRange) =
+    interface ITypeProviderDiagnosticsContext with
+        member _.ReportDiagnostic (staticParameterName, rangeInParameterIfString, message, severity) =
+            let m =
+                match staticArgs |> Array.tryFind (fun x -> x.Name = staticParameterName) with
+                | Some x when x.ValueRange <> range0 ->
+                    match x.ValueRangeAdjusted, rangeInParameterIfString with
+                    | Some m, Some (startPosition, endPosition) when startPosition < endPosition && startPosition >= 0 ->
+                        // todo highlighted range spanning multiple lines :/
+                        mkFileIndexRange m.FileIndex (mkPos m.StartLine (m.StartColumn + startPosition)) (mkPos m.EndLine (m.StartColumn + endPosition))
```

Comment:
I hate this, but I didn't find any facility that would allow me to slice `range` by a certain amount of characters. It's brittle and fails when the diagnostic is supposed to span multiple lines. Don't know if it can be improved without reading the original document to see where the line breaks are.

Reply:
The closest I can see is how it is done for mkFirstLineOfFile , which also does need to scan the entire file.
(if we wanted to avoid it at any cost, we would need to pass an array of line widths I guess)

Reply:
Don did mention that reconstructing the source code range from the logical range is non-trivial, so we could accept this as one of the limitations.